test/specs/lib/FocusZone-test.tsx

```diff
@@ -1110,4 +1110,71 @@ describe('FocusZone', () => {
     expect(inputA.tabIndex).toBe(-1)
     expect(buttonB.tabIndex).toBe(0)
   })
+
+  it('should force focus to first focusable element when FocusZone container receives focus and shouldFocusFirstElementWhenReceivedFocus is set to "true"', () => {
+    const isInnerZoneKeystroke = (e: React.KeyboardEvent<HTMLElement>): boolean =>
+      keyboardKey.getCode(e) === keyboardKey.Enter
+    const isFocusableProperty = { [IS_FOCUSABLE_ATTRIBUTE]: true }
+
+    const component = ReactTestUtils.renderIntoDocument(
+      <div {...{ onFocusCapture: onFocus }}>
+        <FocusZone
+          direction={FocusZoneDirection.horizontal}
+          shouldEnterInnerZone={isInnerZoneKeystroke}
+        >
+          <button className="a">a</button>
+          <FocusZone
+            direction={FocusZoneDirection.horizontal}
+            className="b"
+            shouldFocusFirstElementWhenReceivedFocus={true}
+            {...isFocusableProperty}
+          >
+            <button className="bsub">bsub</button>
+          </FocusZone>
+        </FocusZone>
+      </div>,
+    )
+
+    const focusZone = ReactDOM.findDOMNode(component as React.ReactInstance)!.firstChild as Element
+
+    const buttonA = focusZone.querySelector('.a') as HTMLElement
+    const divB = focusZone.querySelector('.b') as HTMLElement
+
+    const buttonB = focusZone.querySelector('.bsub') as HTMLElement
+
+    setupElement(buttonA, {
+      clientRect: {
+        top: 0,
+        bottom: 20,
+        left: 0,
+        right: 20,
+      },
+    })
+
+    setupElement(divB, {
+      clientRect: {
+        top: 0,
+        bottom: 20,
+        left: 20,
+        right: 40,
+      },
+    })
+
+    setupElement(buttonB, {
+      clientRect: {
+        top: 5,
+        bottom: 15,
+        left: 25,
+        right: 35,
+      },
+    })
+
+    // Focus the first button.
+    ReactTestUtils.Simulate.focus(buttonA)
+    expect(lastFocusedElement).toBe(buttonA)
+
+    ReactTestUtils.Simulate.keyDown(focusZone, { which: keyboardKey.ArrowRight })
+    // Focus goes to FocusZone container, which forces focus to first focusable element - buttonB
+    expect(lastFocusedElement).toBe(buttonB)
```

Comment:
to be sure, what would be the default behavior if this `focusFirstElement` prop won't be set? To me it seems that the following DOM tree is rendered:

```html
<div data='top focus zone container'>
  <button id='a' />

  <div data='focus zone container of b'>
     <button id='b' />
  </div>
</div>
```

It seems that even without this prop (`focusOnFirstElement` for `b` focus zone) behavior will be the same as test suggests - most probably missing something, could you clarify please?

Reply:
Current ```FocusZone``` implementation constraints propagating focuses across nested focus zones. If you need to go inside some nested focus zone, you have to use a prop ```shouldEnterInnerZone```, for e.g.
```js
shouldEnterInnerZone = (e:): boolean =>
      keyboardKey.getCode(e) === keyboardKey.Enter
```
so specify some different key which you'll use to go inside focus zone, so focus elements inside it.

The following tree which is rendered is like you mentioned, but the only difference that inner FocusZone has ```data-is-focusable="true"``` attribute.

```js
<div data='top focus zone container'>
  <button id='a' />

  <div data='focus zone container of b'
          data-is-focusable="true"
          shouldEnterInnerZone="Enter key">
     <button id='b' />
  </div>

   <button id='c' />
</div>
```

Here is a user case without ```focusFirstElement``` : you have a top focus zone container, across which you navigate with arrows Left / Right.
1. you press```->```
2. button ```b``` gets focus
3. you press ```->```
4. Sub focus zone ```focus zone container of b``` gets focus
    4.1 at this point you can press either ```Enter``` to go inside this FocusZone or ```->``` to focus the next element button ```c```
5. you press ```Enter``` 
6. Focus goes inside sub focuszone, so now you can navigate to button ```b```


Case with ```focusFirstElement``` set to ```true```
1. you press```->```
2. button ```b``` gets focus
3. you press ```->```
4. Sub focus zone ```focus zone container of b``` gets focus and forces focus to first element
5.  so button ```b``` gets focus
6. you press ```->``` and you go out of this sub focus zone (in case it doesn't handle navigation with arrow right/left) 
7. button ```c``` get focus

Please, let me know if it got more clear fo you know

Reply:
> 2. button `b` gets focus

you've meant "button `a`" in all the cases described, right? If so, the explanation is clear, thanks!